lib/stdlib/src/uri_string.erl

```diff
@@ -0,0 +1,2067 @@
+    case is_hex_digit(C0) andalso is_hex_digit(C1) of
+        true ->
+            Int = ?HEX2DEC(C0)*16+?HEX2DEC(C1),
+            transcode_pct(Rest, Acc, <<B/binary, Int>>, InEncoding, OutEncoding);
+        false -> throw({error, invalid_percent_encoding,L})
+    end;
+transcode_pct([_C|_Rest] = L, Acc, B, InEncoding, OutEncoding) ->
+    OutBinary = convert_binary(B, InEncoding, OutEncoding),
+    PctEncUtf8 = percent_encode_segment(OutBinary),
+    Out = lists:reverse(convert_list(PctEncUtf8, utf8)),
+    transcode(L, Out ++ Acc, [], InEncoding, OutEncoding);
+transcode_pct([], Acc, B, InEncoding, OutEncoding) ->
+    OutBinary = convert_binary(B, InEncoding, OutEncoding),
+    PctEncUtf8 = percent_encode_segment(OutBinary),
+    Out = convert_list(PctEncUtf8, utf8),
+    lists:reverse(Acc) ++ Out.
+
+
+%% Convert to binary
+convert_binary(Binary, InEncoding, OutEncoding) ->
+    case unicode:characters_to_binary(Binary, InEncoding, OutEncoding) of
+        {error, _List, RestData} ->
+            throw({error, invalid_input, RestData});
+        {incomplete, _List, RestData} ->
+            throw({error, invalid_input, RestData});
+        Result ->
+            Result
+    end.
+
+
+%% Convert to list
+convert_list(Binary, InEncoding) ->
+    case unicode:characters_to_list(Binary, InEncoding) of
+        {error, _List, RestData} ->
+            throw({error, invalid_input, RestData});
+        {incomplete, _List, RestData} ->
+            throw({error, invalid_input, RestData});
+        Result ->
+            Result
+    end.
+
+
+%% Flatten input list
+flatten_list([], _) ->
+    [];
+flatten_list(L, InEnc) ->
+    flatten_list(L, InEnc, []).
+%%
+flatten_list([H|T], InEnc, Acc) when is_binary(H) ->
+    L = convert_list(H, InEnc),
+    flatten_list(T, InEnc, lists:reverse(L) ++ Acc);
+flatten_list([H|T], InEnc, Acc) when is_list(H) ->
+    flatten_list(H ++ T, InEnc, Acc);
+flatten_list([H|T], InEnc, Acc) ->
+    flatten_list(T, InEnc, [H|Acc]);
+flatten_list([], _InEnc, Acc) ->
+    lists:reverse(Acc);
+flatten_list(Arg, _, _) ->
+    throw({error, invalid_input, Arg}).
+
+
+percent_encode_segment(Segment) ->
+    percent_encode_binary(Segment, <<>>).
+
+
+%%-------------------------------------------------------------------------
+%% Helper functions for compose_query
+%%-------------------------------------------------------------------------
+
+%% Returns separator to be used between key-value pairs
+get_separator(_, L) when length(L) =:= 0 ->
+    <<>>;
+get_separator([], _L) ->
+    <<"&amp;">>;
+get_separator([{separator, amp}], _L) ->
+    <<"&">>;
+get_separator([{separator, escaped_amp}], _L) ->
+    <<"&amp;">>;
+get_separator([{separator, semicolon}], _L) ->
+    <<";">>.
+
+
+%% Form-urlencode input based on RFC 1866 [8.2.1]
+form_urlencode(Cs) when is_list(Cs) ->
+    B = convert_binary(Cs, utf8, utf8),
+    form_urlencode(B, <<>>);
+form_urlencode(Cs) ->
+    form_urlencode(Cs, <<>>
```

Comment:
Reminder that we think this should eventually be merged with inet_db:tolower and made into an API function probably
in inet.